File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.cpp
// class Solution {
// public:
//     int lengthOfLongestSubstring(string s) {
//         int maxLength = 0;

//         unordered_map<char, int> mp;

//         int left = 0;
//         int right = 0;

//         int n = s.length();

//         int count = 0;

//         while(right < n) {
//             if(mp.find(s[right]) == mp.end()) {
//                 mp[s[right]]++;
//                 count++;
//                 right++;
//             } else {
//                 count = 0;
//                 left++;
//                 right = left;
//                 mp.clear();
//             }
//             maxLength = max(maxLength, count);
//         }

//         return maxLength;
//     }
// };


class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int maxLength = 0;

        int n = s.length();

        vector<int> mp(256, -1);

        int l = 0;
        int r = 0; 

        while(r < n) {
            if(mp[s[r]] != -1) {
                if(mp[s[r]] >= l) {
                    l = mp[s[r]] + 1;
                }
            }
            int length = r - l + 1;
            maxLength = max(maxLength, length);

            mp[s[r]] = r;
            r++;
        }
                                   
        return maxLength;
    }
};

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.java
// class Solution {
//     public int lengthOfLongestSubstring(String s) {
//         int result = 0;
//         int n = s.length();

//         for(int i = 0; i < n; i++) {
//             HashSet<Character> set = new HashSet<>();
//             int count = 0;
//             int j = i;
//             while(j < n && !set.contains(s.charAt(j))) {
//                 set.add(s.charAt(j));
//                 count++;
//                 j++;
//             }

//             result = Math.max(count, result);
//         }

//         return result;
//     }
// }


class Solution {
    public int lengthOfLongestSubstring(String s) {
        int result = 0;
        int n = s.length();
        HashSet<Character> set = new HashSet<>();
        int j = 0;

        for(int i = 0; i < n; i++) {
            while(set.contains(s.charAt(i))) {
                set.remove(s.charAt(j));
                j++;
            }

            set.add(s.charAt(i));

            result = Math.max(i-j+1, result);
        }

        return result;
    }
}